src/registry.rs

```diff
@@ -365,15 +367,54 @@ impl Unit {
     }
 }
 
+// TODO: Instead of depending on dynamic dispatch for all metrics, we could use
+// static dispatch for the most common ones and only fall back to dynamic
+// dispatch. That said, this needs to be proven performant in a benchmark first.
+/// Super trait representing an abstract Prometheus metric.
+#[cfg(not(feature = "protobuf"))]
+pub trait Metric:
+    crate::encoding::text::EncodeMetric + Send + Sync + std::fmt::Debug + 'static
+{
+}
+
+/// Super trait representing an abstract Prometheus metric.
+#[cfg(feature = "protobuf")]
+pub trait Metric:
+    crate::encoding::text::EncodeMetric
+    + crate::encoding::proto::EncodeMetric
+    + Send
+    + Sync
+    + std::fmt::Debug
+    + 'static
+{
+}
+
+#[cfg(not(feature = "protobuf"))]
+impl<T> Metric for T where
+    T: crate::encoding::text::EncodeMetric + Send + Sync + std::fmt::Debug + 'static
+{
+}
+
+#[cfg(feature = "protobuf")]
+impl<T> Metric for T where
+    T: crate::encoding::text::EncodeMetric
+        + crate::encoding::proto::EncodeMetric
+        + Send
+        + Sync
+        + std::fmt::Debug
+        + 'static
+{
+}
```

Comment:
This is a bit hacky but I guess it works? :D

Reply:
Agreed. Though I am not sure how to do it differently.

Reply:
Features are supposed to be additive, this is not additive.

Reply:
Functionality-wise for the end user, it is additive. There is more discussion further down in how else we could solve this!

Reply:
If I care about only text encoding and my metric type implements only `crate::encoding::text::EncodeMetric`, and some dependency enables the `protobuf`, then my type won't implement `Metric` anymore.

Reply:
It will. The one defined on line 382.

There will always be a `Metric` trait, just with different supertraits basd on the activated feature.

I can't think of a situation in which this breaks something but I agree that it is not clean.

Reply:
Okay, yeah. It is technically breaking. But the set of prometheus metrics is fixed right? Why would you implement the trait yourself? What am I missing?

Reply:
I've already implemented [my own `Family`](https://github.com/nox/prometools/blob/11e2ce6248424b73643c0c9ec6f0e0c639abab54/src/serde/mod.rs#L19-L82) and [my own `Counter`](https://github.com/nox/prometools/blob/11e2ce6248424b73643c0c9ec6f0e0c639abab54/src/nonstandard.rs#L16-L19).

Reply:
Thanks for raising this @nox.

I think the way to go is a single `EncodeMetric` trait which then allows `encoding:text` and `encoding::protobuf` to traverse a metric via the visitor patter as suggested below by @thomaseizinger.